db/commitlog/commitlog.cc

```diff
@@ -1841,6 +1841,64 @@ future<db::rp_handle> db::commitlog::add_entry(const cf_id_type& id, const commi
     });
 }
 
+future<std::vector<db::rp_handle>> 
+db::commitlog::add_entries(std::vector<commitlog_entry_writer> entry_writers, db::timeout_clock::time_point timeout) {
+    class cl_entries_writer final : public entry_writer {
+        std::vector<commitlog_entry_writer> _writers;
+    public:
+        std::vector<rp_handle> res;
+
+        cl_entries_writer(force_sync sync, std::vector<commitlog_entry_writer> entry_writers)
+            : entry_writer(sync, entry_writers.size()), _writers(std::move(entry_writers))
+        {
+            res.reserve(_writers.size());
+        }
+        const cf_id_type& id(size_t i) const override {
+            return _writers.at(i).schema()->id();
+        }
+        size_t size(segment& seg) override {
+            size_t res = 0;
+            for (auto i = _writers.begin(), e = _writers.end(); i != e; ++i) {
+                auto known = seg.is_schema_version_known(i->schema());
+                if (!known) {
+                    for (auto j = _writers.begin(); j != i; ++j) {
+                        if (j->schema() == i->schema()) {
+                            known = true;
+                            break;
+                        }
+                    }
+                }
+                i->set_with_schema(!known);
+                res += i->size();
+            }
+            return res;
+        }
+        size_t size(segment& seg, size_t i) override {
+            return _writers.at(i).size(); // we have already set schema known/unknown
```

Comment:
`size(segment&)` has to be called before `size(segment&, size_t)` can be called? Seams kind of fragile.

Reply:
Given that it is 100% internal dispatching to known code I don't know if I think it really matters. Can of course set a flag just because, but it seems more reasonable to just add documentation to the internal abstract type that `size(segment&)` always is called first (total estimate), and `size(segment&, size_t)` _only_ is called if num_elem > 1.

Reply:
The interface is indeed tightly coupled with its only user. May be documentation suffice. To be honest three different size() functions confuse me. They _are_ documented, but documentation assumes knowledge of the code internals.